pkg/instancetype/upgrade.go

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+package instancetype
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	instancetypeapi "kubevirt.io/api/instancetype"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+)
+
+type Upgrader interface {
+	Upgrade(vm *virtv1.VirtualMachine) error
+}
+
+func (m *InstancetypeMethods) Upgrade(vm *virtv1.VirtualMachine) error {
+	// Skip if the VirtualMachine doesn't reference an instance type or preference
+	if vm.Spec.Instancetype == nil && vm.Spec.Preference == nil {
+		return nil
+	}
+
+	var (
+		err                        error
+		originalInstancetypeCRName string
+		originalPreferenceCRName   string
+		newInstancetypeCR          *appsv1.ControllerRevision
+		newPreferenceCR            *appsv1.ControllerRevision
+		patchPayload               []byte
+	)
+
+	if vm.Spec.Instancetype != nil && vm.Spec.Instancetype.RevisionName != "" {
+		originalInstancetypeCRName = vm.Spec.Instancetype.RevisionName
+		newInstancetypeCR, patchPayload, err = m.upgradeControllerRevision(vm, vm.Spec.Instancetype.RevisionName, "/spec/instancetype/revisionName")
+		if err != nil {
+			return err
+		}
+	}
+
+	if vm.Spec.Preference != nil && vm.Spec.Preference.RevisionName != "" {
+		originalPreferenceCRName = vm.Spec.Preference.RevisionName
+		var preferencePatchPayload []byte
+		newPreferenceCR, preferencePatchPayload, err = m.upgradeControllerRevision(vm, vm.Spec.Preference.RevisionName, "/spec/preference/revisionName")
+		if err != nil {
+			return err
+		}
+		patchPayload = append(patchPayload, preferencePatchPayload...)
+	}
+
+	// TODO - Refactor GenerateRevisionNamePatch and reuse it here to generate the patch
+	if patchPayload == nil {
+		return nil
+	}
+
+	if _, err := m.Clientset.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchPayload, &metav1.PatchOptions{}); err != nil {
+		return err
+	}
+
+	if newInstancetypeCR != nil {
+		if err := m.Clientset.AppsV1().ControllerRevisions(vm.Namespace).Delete(context.Background(), originalInstancetypeCRName, metav1.DeleteOptions{}); err != nil {
+			log.Log.Object(vm).Reason(err).Error("ignoring failure to delete ControllerRevision during stashed instance type object upgrade")
+		}
+	}
+
+	if newPreferenceCR != nil {
+		if err := m.Clientset.AppsV1().ControllerRevisions(vm.Namespace).Delete(context.Background(), originalPreferenceCRName, metav1.DeleteOptions{}); err != nil {
+			log.Log.Object(vm).Reason(err).Error("ignoring failure to delete ControllerRevision during stashed preference object upgrade")
+		}
+	}
+
+	log.Log.Object(vm).Info("instancetype.kubevirt.io ControllerRevisions upgrade successful")
```

Comment:
This seems to be very verbose, it could be emitted a lot of times even if no upgrade was necessary? (no-op above) I'd suggest to raise the verbosity level.

Reply:
no-op's should have already returned above before we call patch so this shouldn't be verbose no.

Reply:
ACK